pkg/controller/garbagecollector/graph.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type objectReference struct {
+	metav1.OwnerReference
+	// This is needed by the dynamic client
+	Namespace string
+}
+
+func (s objectReference) String() string {
+	return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
+}
+
+// The single-threaded GraphBuilder.processEvent() is the sole writer of the
+// nodes. The multi-threaded GarbageCollector.processItem() reads the nodes.
+type node struct {
+	identity objectReference
+	// dependents will be read by the orphan() routine, we need to protect it with a lock.
+	dependentsLock sync.RWMutex
```

Comment:
Talked in person, the concern is that if different fields are accessed in sequence using multiple locks, the caller might get an inconsistent view. There are two places these fields are accessed in sequence, but both are innocuous.
1. in `attemptToDeleteItem`,
```go
// isBeingDeleted and isDeletingDependents hold two different locks
if item.isBeingDeleted() && !item.isDeletingDependents() {
   ...
}
```
We never change beingDeleted from true to false, so it's fine.

2. also in `attemptToDeleteItem`,
```go
// acquire and then release deletingDependents lock
if item.isDeletingDependents() {
   ...
   // acquire and then release dependentsLock 
   blockingDependents := item.blockingDependents()
```
Same story here, we never change isDeletingDependents from true to false, so it's ok.